torchsde/_core/methods/milstein.py

```diff
@@ -12,30 +12,67 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS
+import math
+
+import torch
+
+from ...settings import SDE_TYPES, NOISE_TYPES, LEVY_AREA_APPROXIMATIONS, METHOD_OPTIONS as opt
 
 from .. import base_solver
 
 
-class Milstein(base_solver.BaseSDESolver):
+class BaseMilstein(base_solver.BaseSDESolver):
     strong_order = 1.0
     weak_order = 1.0
-    sde_type = SDE_TYPES.ito
     noise_types = (NOISE_TYPES.additive, NOISE_TYPES.diagonal, NOISE_TYPES.scalar)
     levy_area_approximations = LEVY_AREA_APPROXIMATIONS.all()
 
+    def v_term(self, I_k, dt):
+        pass
+
     def step(self, t0, y0, dt):
         assert dt > 0, 'Underflow in dt {}'.format(dt)
 
-        I_k = self.bm(t0, t0 + dt)
-        v = [delta_bm_ ** 2. - dt for delta_bm_ in I_k]
+        t1 = t0 + dt
+
+        I_k = self.bm(t0, t1)
+        v = self.v_term(I_k, dt)
 
         f_eval = self.sde.f(t0, y0)
-        g_prod_eval = self.sde.g_prod(t0, y0, I_k)
-        gdg_prod_eval = self.sde.gdg_prod(t0, y0, v)
+        g_eval = self.sde.g(t0, y0)
+        g_prod_eval = self.sde.prod(g_eval, I_k)
+
+        if opt.grad_free in self.options and self.options[opt.grad_free]:
+            g_prod_eval_v = self.sde.prod(g_eval, v)
+            sqrt_dt = torch.sqrt(dt) if isinstance(dt, torch.Tensor) else math.sqrt(dt)
+            y0_prime = [
+                y0_ + dt * f_eval_ + g_eval_ * sqrt_dt
+                for y0_, f_eval_, g_eval_ in zip(y0, f_eval, g_eval)
+            ]            
+            g_prod_eval_prime = self.sde.g_prod(t1, y0_prime, v)
+            gdg_prod_eval = [
+                (g_prod_eval_prime_ - g_prod_eval_v_) / sqrt_dt
+                for g_prod_eval_prime_, g_prod_eval_v_ in zip(g_prod_eval_prime, g_prod_eval_v)
+            ]
+        else:
+            gdg_prod_eval = self.sde.gdg_prod(t0, y0, v)
+        
         y1 = [
             y0_i + f_eval_i * dt + g_prod_eval_i + .5 * gdg_prod_eval_i
             for y0_i, f_eval_i, g_prod_eval_i, gdg_prod_eval_i in zip(y0, f_eval, g_prod_eval, gdg_prod_eval)
         ]
-        t1 = t0 + dt
         return t1, y1
+
+
+class MilsteinIto(BaseMilstein):
+    sde_type = SDE_TYPES.ito
+    
+    def v_term(self, I_k, dt):
+        return [delta_bm_ ** 2. - dt for delta_bm_ in I_k]
```

Comment:
Can this (and the line below) be `** 2` rather than `** 2.`?
I suspect (but do not know) that an integer power will be implemented as the cheap `x**2 = x * x` whilst a floating power will be implemented as the expensive `x ** 2. = exp(2. * log(x))`.
(Happy to be corrected on this if you know better.)

Reply:
I don't know that - changed to `** 2`.